source/common/http/websocket/ws_handler_impl.cc

```diff
@@ -0,0 +1,145 @@
+    headers->addStatic(Headers::get().Status,
+                       std::to_string(enumToInt(Http::Code::ServiceUnavailable)));
+    stream_.encodeHeaders(std::move(headers), true);
+    return;
+  }
+
+  Filter::TcpProxy::commonInitializeUpstreamConnection(cluster);
+}
+
+void WsHandlerImpl::onConnectTimeout() {
+  ENVOY_CONN_LOG(debug, "connect timeout", read_callbacks_->connection());
+  read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_connect_timeout_.inc();
+  upstream_connection_.get()->close(Network::ConnectionCloseType::NoFlush);
+  Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+  headers->addStatic(Headers::get().Status, std::to_string(enumToInt(Http::Code::GatewayTimeout)));
+  stream_.encodeHeaders(std::move(headers), true);
+}
+
+void WsHandlerImpl::onUpstreamEvent(uint32_t event) {
+  if (event & Network::ConnectionEvent::RemoteClose) {
+    read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_destroy_remote_.inc();
+  }
+
+  if (event & Network::ConnectionEvent::LocalClose) {
+    read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_destroy_local_.inc();
+  }
+
+  if (event & Network::ConnectionEvent::RemoteClose) {
+    if (connect_timeout_timer_) {
+      read_callbacks_->upstreamHost()->cluster().stats().upstream_cx_connect_fail_.inc();
+      read_callbacks_->upstreamHost()->stats().cx_connect_fail_.inc();
+      Http::HeaderMapPtr headers = std::unique_ptr<Http::HeaderMap>(new Http::HeaderMapImpl());
+      headers->addStatic(Headers::get().Status, std::to_string(enumToInt(Http::Code::BadGateway)));
+      stream_.encodeHeaders(std::move(headers), true);
+    }
+
+    read_callbacks_->connection().close(Network::ConnectionCloseType::FlushWrite);
+  } else if (event & Network::ConnectionEvent::Connected) {
+    connect_timespan_->complete();
+
+    // Wrap upstream connection in HTTP Connection, so that we can
+    // re-use the HTTP1 codec to send upgrade headers to upstream
+    // host.
+
+    // TODO (rshriram): This is a not the strictest WebSocket
+    // implementation, as we do not really check the response headers
+    // to ensure that the upstream really accepted the upgrade
+    // request.  Doing so requires re-doing bunch of HTTP/1.1 req-resp
+    // pair stuff and that is going to just complicate this code. The
+    // client could technically send a body along with the request.
+    // The server could send a body along with the upgrade response,
+    // or even send a redirect. A simpler way to handle all of this would be
+    // to expose the TCP connection associated with the response object, in encodeHeaders.
+    // Once we see a 101 switching protocols, we could remove the connection from
+    // the connection pool. The current approach is a stop gap solution, where
+    // we put the onus on the user to tell us if a route (and corresponding upstream)
+    // is supposed to allow websocket upgrades or not.
```

Comment:
Yes, user is the person configuring Envoy. And I have the same concern here as well.